numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1200,7 +1200,8 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
             }
         }
         if (res == NULL && !PyErr_Occurred()) {
-            PyErr_SetString(PyExc_ValueError, "No fields found.");
+            /* these dtypes had no fields */
+            return cmp_op == Py_EQ ? Py_True : Py_False;
```

Comment:
This change isn't correct - it doesn't produce a correctly broadcast array. Let's leave this to a future PR - the rest looks great.

Reply:
Fixed by falling through to the string compare case which does the right thing here.